cmd/syncComparator.go

```diff
@@ -156,66 +156,91 @@ func (f *syncDestinationComparator) HasFileChangedSinceLastSyncUsingLocalChecks(
 // This condition is conveyed by passing the 2nd argument (FinalizeAll) as false.
 //
 // This function will also update the folder metaData, if changed which is in case of finalizeAll true and delete the flderMap from the ObjectIndexerMap.
-func (f *syncDestinationComparator) FinalizeTargetDirectory(folderMap *objectIndexer, finalizeAll bool, lcFolderName string) {
+func (f *syncDestinationComparator) FinalizeTargetDirectory(lcFolderName string, finalizeAll bool) {
 	var size int64
+
+	// No need to check folderPresent, as it's already checked by caller.
+	folderMap, _ := f.sourceFolderIndex.folderMap[lcFolderName]
+
```

Comment:
at least let's panic if it's not found, accompanied by a comment why we are asserting

also, what happens if lcFolderName is "" i.e. it's called for root folder. I see that you are depending on the caller not calling with that.
Also, same holds true for the above assertion that you make. Instead, it would be better to do all this here.

I suggest do the following here

change arg lcFolderName to relativePath, indicating that it's a relative path of the directory.
then we do all the checks here, and remove these from the caller

        // Root folder is represented as "." in indexer map.
        if relativePath == "" {
            relativePath = "."
        }

        folderMap, folderPresent := f.sourceFolderIndex.folderMap[relativePath]

        if !folderPresent {
                panic(fmt.Sprintf("Folder with relativePath[%s] not present in ObjectIndexerMap", relativePath))
        }

        fmt.Printf("Finalizing directory %s (FinalizeAll=%v)\n", relativePath, finalizeAll)

Reply:
Agreed, i thought of the same. But my thought now this Finalizer getting long. Other reason for that, its single thread and nobody else delete the map except the processIfNecessary.